File: python/stack/maximum-frequency-stack.py
import heapq


class FreqStack:

    def __init__(self):
        self.heap = []
        self.count = defaultdict(lambda: 0)
        self.time = 0

    def push(self, val: int) -> None:
        self.count[val] += 1
        heapq.heappush(self.heap, (-self.count[val], self.time, val))
        self.time -= 1

    def pop(self) -> int:
        count, time, popped = heapq.heappop(self.heap)
        self.count[popped] -= 1
        return popped

# Your FreqStack object will be instantiated and called as such:
# obj = FreqStack()
# obj.push(val)
# param_2 = obj.pop()

File: c++/stack/maximum-frequency-stack.cc
class FreqStack {
public:
    FreqStack() {
        time_ = 0;
    }

    void push(int val) {
        count_[val]++;
        heap_.push(std::make_tuple(count_[val], time_, val));
        time_++;
    }

    int pop() {
        std::tuple<int, int, int> element = heap_.top();
        count_[std::get<2>(element)]--;
        heap_.pop();
        return std::get<2>(element);
    }

private:
    std::priority_queue<std::tuple<int, int, int>> heap_;
    std::unordered_map<int, int> count_;
    int time_;
};

/**
 * Your FreqStack object will be instantiated and called as such:
 * FreqStack* obj = new FreqStack();
 * obj->push(val);
 * int param_2 = obj->pop();
 */
